torch/_export/utils.py

```diff
@@ -1350,3 +1350,69 @@ def default_flatten_fn_spec(obj, spec) -> list[Any]:
 def deregister_module_as_pytree_input_node(cls: type[torch.nn.Module]) -> None:
     _deregister_pytree_node(cls)
     _deregister_pytree_flatten_spec(cls)
+
+
+def _enforce_placeholder_order_pass(
+    gm: torch.fx.GraphModule, graph_signature: "ExportGraphSignature"
+):
+    # Enforce the order of placeholder nodes in the graph module to be:
+    # token -> parameter -> buffer (persistent) -> buffer (non-persistent)
+    # -> tensor_constant -> custom_obj -> user_inputs
+    input_nodes = [node for node in gm.graph.nodes if node.op == "placeholder"]
+    if len(input_nodes) <= 1:
+        return
+
+    reordered_input_nodes = []
+    reordered_input_specs = []
+
+    for input_kind in [
+        InputKind.TOKEN,
+        InputKind.PARAMETER,
+        InputKind.BUFFER,
+        InputKind.CONSTANT_TENSOR,
+        InputKind.CUSTOM_OBJ,
+        InputKind.USER_INPUT,
+    ]:
+        if input_kind == InputKind.BUFFER:
+            # Enforce that persistent buffers always come before non-persistent ones
+            for input_spec, input_node in zip(graph_signature.input_specs, input_nodes):
+                if input_spec.kind == input_kind and input_spec.persistent:
+                    reordered_input_nodes.append(input_node)
+                    reordered_input_specs.append(input_spec)
+            for input_spec, input_node in zip(graph_signature.input_specs, input_nodes):
+                if input_spec.kind == input_kind and not input_spec.persistent:
+                    reordered_input_nodes.append(input_node)
+                    reordered_input_specs.append(input_spec)
+        else:
+            for input_spec, input_node in zip(graph_signature.input_specs, input_nodes):
+                if input_spec.kind == input_kind:
+                    reordered_input_nodes.append(input_node)
+                    reordered_input_specs.append(input_spec)
+
+    assert len(reordered_input_specs) == len(graph_signature.input_specs)
+    assert len(reordered_input_nodes) == len(input_nodes)
+
+    with gm.graph.inserting_before(input_nodes[0]):
```

Comment:
I believe just doing `gm.graph.inserting_before()` will put the nodes at the very top